guidance/hybridpath_guidance/hybridpath_guidance/hybridpath.py

```diff
@@ -0,0 +1,243 @@
+                # self.Path['coeff']['b_der'].append(b_poly)
+
+    def plot_path(self):
+        if self.plot_handle is not None:
+            e = np.arange(0, 1, 0.01)
+            plt.figure(self.plot_handle)
+            plt.clf()
+            for j in range(self.Path['NumSubpaths']):
+                X,Y = np.zeros_like(e), np.zeros_like(e)
+                for k in range(self.ord + 1):
+                    X += self.Path['coeff']['a'][j][k] * e**k
+                    Y += self.Path['coeff']['b'][j][k] * e**k
+                if j == 0:
+                    plt.plot(Y,X,'b', linewidth = 1.25, label = 'ASV Reference Path')
+                else:
+                    plt.plot(Y,X,'b', linewidth = 1.25)
+            
+            plt.plot(self.WP[:, 1], self.WP[:, 0], 'r*', linewidth = 1.5, label = 'Waypoints')
+            plt.xlabel('y')
+            plt.ylabel('x')
+            plt.axis('equal')
+            minX, maxX = np.min(self.WP[:, 1]), np.max(self.WP[:, 1])
+            minY, maxY = np.min(self.WP[:, 0]), np.max(self.WP[:, 0])
+            plt.axis([1.1*minX - 0.1*maxX, 1.1*maxX - 0.1*minX, 1.1*minY - 0.1*maxY, 1.1*maxY - 0.1*minY])
+            plt.grid()
+            plt.legend()
+            plt.title('Path')
+            plt.show()
+
+            
+class HybridPathSignals:
+    def __init__(self, path, s):
+        self.path = path
+        self.s = self._clamp_s(s, path['NumSubpaths'])
+        self.ord = path['Order']
+        self.pd = None
+        self.pd_der = []
+        self.pd, self.pd_der = self.get_signals()
+        self.psi, self.psi_der, self.psi_dder = self.get_heading()
+        
+    
+    def _clamp_s(self, s, num_subpaths):
+        """
+        Ensure s is within the valid range.
+        """
+        if s < 0:
+            return 0.0
+        elif s >= num_subpaths:
+            return num_subpaths - 1e-3 # A small epsilon
+        return s
+    
+    def get_signals(self):
+        """
+        Compute the location and derivative at s.
+        """
+        idx = int(self.s) + 1
+        t = self.s - (idx - 1)
+        xd, yd = 0, 0
+        # Compute position
+        for k in range(self.ord + 1):
+            xd += self.path['coeff']['a'][idx - 1][k] * t**k
+            yd += self.path['coeff']['b'][idx - 1][k] * t**k
+        self.pd = [xd, yd]
+
+        # Compute derivatives
+        for k in range(1, self.ord + 1): ##
+            xd_der, yd_der = 0, 0
+            # print('Index number: ', idx, ' was OK! Max is: ', self.path['NumSubpaths'])
+            # print('k: ', k, ' was OK! Max is: ', len(self.path['coeff']['a_der']))
+            a_vec = self.path['coeff']['a_der'][k - 1][idx - 1]
+            b_vec = self.path['coeff']['b_der'][k - 1][idx - 1]
+            for j in range(len(a_vec)):
+                xd_der += a_vec[j] * t**j
+                yd_der += b_vec[j] * t**j
+            self.pd_der.append([xd_der, yd_der])
+
+        return self.pd, self.pd_der
+    
+    def get_heading(self):
```

Comment:
method is specified as `get_heading()` but also returns derivatives. I'd suggest splitting them up and providing a single comment at the top specifying what equation is used for each method, since you can't really avoid the mess of indexing with the current structure of self.pd_der